File: net/Link.Message.Client/content/complex/linear/GridContent.cs
using System.Collections.Generic;
using Link.Message.Client.utils;
using Newtonsoft.Json;

namespace Link.Message.Client.content.complex.linear
{
	public class GridContent : IStringOrObjectSerializable
	{
		protected internal IList<Column> columns = new List<Column>();
		protected internal IList<string[]> datas = new List<string[]>();
		protected internal int columnCount;

		public GridContent(int columnCount)
		{
			this.columnCount = columnCount;
		}

        [JsonProperty("column")]
		public virtual IList<Column> Columns
		{
			get
			{
				return columns;
			}
		}

        [JsonProperty("data")]
		public virtual IList<string[]> Datas
		{
			get
			{
				return datas;
			}
		}

        [JsonIgnore]
		public virtual int ColumnCount
		{
			get
			{
				return columnCount;
			}
		}

		public virtual void AddColumn(Column column)
		{
			Guard.GuardReqiredObject(column, "column must be set value.");

			if (columns.Count == columnCount)
			{
				Guard.ThrowIllegalArgumentException("grid columns size already equal column count.");
			}

			columns.Add(column);
		}

		public virtual void AddData(string[] data)
		{
			Guard.GuardReqiredArray(data, "data must be set value.");

			if (data.Length != columnCount)
			{
				Guard.ThrowIllegalArgumentException("data array's size must equal column count.");
			}

			datas.Add(data);
		}
	}
}

File: java/link-message-client/src/main/java/link/message/client/content/complex/linear/GridContent.java
package link.message.client.content.complex.linear;

import java.util.ArrayList;
import java.util.List;

import link.message.client.StringOrObjectSerializable;
import link.message.client.utils.Guard;

import com.alibaba.fastjson.annotation.JSONField;

public class GridContent implements StringOrObjectSerializable {
	@JSONField(name="column")
	protected List<Column> columns = new ArrayList<Column>();
	@JSONField(name="data")
	protected List<String[]> datas  = new ArrayList<String[]>();
	@JSONField(serialize=false)
	protected int columnCount;
	
	public GridContent(int columnCount) {
		this.columnCount = columnCount;
	}

	public List<Column> getColumns() {
		return columns;
	}

	public List<String[]> getDatas() {
		return datas;
	}
	
	public int getColumnCount() {
		return columnCount;
	}

	public void addColumn(Column column) {
		Guard.guardReqiredObject(column, "column must be set value.");
		
		if (columns.size() == columnCount) {
			Guard.throwIllegalArgumentException("grid columns size already equal column count.");
		}
		
		columns.add(column);
	}
	
	public void addData(String[] data) {
		Guard.guardReqiredArray(data, "data must be set value.");
		
		if (data.length != columnCount) {
			Guard.throwIllegalArgumentException("data array's size must equal column count.");
		}
		
		datas.add(data);
	}
}